### PR TITLE
Tier2 - CephFS-NFS - CEPH-83575097

### PR DESCRIPTION
**Pull request by hkadam134**
Tier 2 - Automation - CEPH-83575097 - Create an NFS HA service with 2 or more daemons on a non-default port and test if the ingress / haproxy /keepalived daemons are running and the failover happens as expected.

Logs : http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-WGNKO8/

# Description

Please include Automation development guidelines. Source of Test case - New Feature/Regression Test/Close loop of customer BZs
<details>

<summary>click to expand checklist</summary>

- [ ] Create a test case in Polarion reviewed and approved.
- [ ] Create a design/automation approach doc. Optional for tests with similar tests already automated.
- [ ] Review the automation design
- [ ] Implement the test script and perform test runs
- [ ] Submit PR for code review and approve
- [ ] Update Polarion Test with Automation script details and update automation fields
- [ ] If automation is part of Close loop, update BZ flag qe-test_coverage “+” and link Polarion test
</details>

### CODE REVIEW

File: tests/cephfs/cephfs_nfs/test_cephfs_nfs_ha_deployment_with_non_default_port.py [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+from ceph.parallel import parallel
+from tests.cephfs.cephfs_utilsV1 import FsUtils as FsUtilsV1
+from utility.log import Log
+
+log = Log(__name__)
+
+
+def run(ceph_cluster, **kw):
+    """
+    CEPH-83575097 - Create an NFS HA service with 2 or more daemons on a non default port
```
Comment (sumabai) [inline tests/cephfs/cephfs_nfs/test_cephfs_nfs_ha_deployment_with_non_default_port.py:19]: Please add this automation script path in polarion TC field - Automation script

File: tests/cephfs/cephfs_nfs/test_cephfs_nfs_ha_deployment_with_non_default_port.py [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+        log.info(f"Running cephfs {tc} test case")
+        config = kw["config"]
+        build = config.get("build", config.get("rhbuild"))
+        fs_util_v1 = FsUtilsV1(ceph_cluster)
+        clients = ceph_cluster.get_ceph_objects("client")
+        client1 = clients[0]
+        fs_util_v1.prepare_clients(clients, build)
+        fs_util_v1.auth_list(clients)
+        nfs_name = "cephfsnfs"
+        virtual_ip = "10.8.128.100"
```
Comment (sumabai) [inline tests/cephfs/cephfs_nfs/test_cephfs_nfs_ha_deployment_with_non_default_port.py:46]: should the IP addr be always hard-coded?

File: tests/cephfs/cephfs_nfs/test_cephfs_nfs_ha_deployment_with_non_default_port.py [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+        if not clients:
+            log.info(
+                f"This test requires minimum 1 client nodes.This has only {len(clients)} clients"
+            )
+            return 1
+
+        client1.exec_command(sudo=True, cmd="ceph mgr module enable nfs")
+        client1 = clients[0]
+        nfs_servers = ceph_cluster.get_ceph_objects("nfs")
+        nfs_name = "cephnfs"
```
Comment (sumabai) [inline tests/cephfs/cephfs_nfs/test_cephfs_nfs_ha_deployment_with_non_default_port.py:60]: can the nfs name be 'cephfs-nfs' to look similar what is used in existing nfs tests?

**Review by sumabai — CHANGES_REQUESTED** [review]

**Comment by neha-gangadhar** [comment]
@hkadam134 "get_deamon_status" added in PR : https://github.com/red-hat-storage/cephci/pull/2965/ can be used to validate all services are up giving more details as to which service is not up in case of such scenarios. Please include the individual services status as well.

File: tests/cephfs/cephfs_nfs/test_cephfs_nfs_ha_deployment_with_non_default_port.py (hunk 2 above)
Comment (hkadam134) [reply tests/cephfs/cephfs_nfs/test_cephfs_nfs_ha_deployment_with_non_default_port.py:46]: Yeah, This is the free IP I requested from the Octo Lab maintainer and will be dedicated to Virtual IP for running nfs-ha suites.
As NFS-HA can be automated only in bare-metal, we do not have any other option to fetch free IPs and allocate it.

File: tests/cephfs/cephfs_nfs/test_cephfs_nfs_ha_deployment_with_non_default_port.py (hunk 3 above)
Comment (hkadam134) [reply tests/cephfs/cephfs_nfs/test_cephfs_nfs_ha_deployment_with_non_default_port.py:60]: Sure

File: tests/cephfs/cephfs_nfs/test_cephfs_nfs_ha_deployment_with_non_default_port.py (hunk 1 above)
Comment (hkadam134) [reply tests/cephfs/cephfs_nfs/test_cephfs_nfs_ha_deployment_with_non_default_port.py:19]: I ll add once all the review is completed.

File: tests/cephfs/cephfs_utilsV1.py [hunk 4]
```diff
@@ -2981,3 +2981,26 @@ def wait_for_service_to_be_in_running(self, client1, node):
         for service in services:
             if service["status"] != 1:
                 raise CommandFailed(f"service : {service['service_name']} is not UP")
+
+    def get_active_nfs(self, nfs_servers, virtual_ip):
```
Comment (neha-gangadhar) [inline tests/cephfs/cephfs_utilsV1.py:2985]: can the functions "get_active_nfs" and "get_active_nfs_server" be moved to utils? I see these 2 functions definition being added in all NFS tests which is not needed. Common modules to be reused and not include in every test.

File: tests/cephfs/cephfs_nfs/test_cephfs_nfs_ha_deployment_with_non_default_port.py [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+        )
+        log.info("Removing NFS Cluster")
+        client1.exec_command(
+            sudo=True,
+            cmd=f"ceph nfs cluster rm {nfs_name}",
+            check_ec=False,
+        )
+
+
+def write_io(client1, mount_dir):
```
Comment (neha-gangadhar) [inline tests/cephfs/cephfs_nfs/test_cephfs_nfs_ha_deployment_with_non_default_port.py:166]: write_io function should in utils and reused in tests.

File: tests/cephfs/cephfs_nfs/test_cephfs_nfs_ha_deployment_with_non_default_port.py (hunk 2 above)
Comment (sumabai) [reply tests/cephfs/cephfs_nfs/test_cephfs_nfs_ha_deployment_with_non_default_port.py:46]: thanks for the info

**Comment by hkadam134** [comment]
> @hkadam134 "get_deamon_status" added in PR : #2965 can be used to validate all services are up giving more details as to which service is not up in case of such scenarios. Please include the individual services status as well.

Done.

**Review by hkadam134 — COMMENTED** [review]
Latest Logs : http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-PD6ENE/

**Review by sumabai — APPROVED** [review]

**Review by neha-gangadhar — APPROVED** [review]

**Pull request merged**